src/supervisor.tex

```diff
@@ -1465,16 +1486,18 @@ \subsection{Virtual Address Translation Process}
 \item If $i>0$ and $pte.ppn[i-1:0]\neq 0$, this is a misaligned superpage;
   stop and raise a page-fault exception corresponding to the original access type.
 
-\item If $pte.a=0$, or if the memory access is a store and $pte.d=0$, either
+\item If $pte.a=0$, or if the original memory access is a store and $pte.d=0$, either
   raise a page-fault exception corresponding to the original access type, or:
   \begin{itemize}
-  \item Set $pte.a$ to 1 and, if the memory access is a store, also set
-    $pte.d$ to 1.
-  \item If this access violates a PMA or PMP check, raise an access-fault exception
-    corresponding to the original access type.
-  \item This update and the loading of $pte$ in step 2 must be atomic; in
-    particular, no intervening store to the PTE may be perceived to have
-    occurred in-between.
+  \item If a store to $pte$ would violate a PMA or PMP check, raise an
+    access-fault exception corresponding to the original access type.
+  \item Perform the following steps atomically:
+    \begin{itemize}
+      \item Compare $pte$ to the value of the PTE at address $a+va.vpn[i]\times \textrm{PTESIZE}$.
```

Comment:
I'm wondering if it would make sense to, if $pte$ is a NAPOT PTE, allow these three steps to be performed against the memory location of the first equivalent PTE.  The motivation would be a reduction in the number of stores which have to be performed.

Reply:
If I understand your suggestion correctly, that would require a change to the address-translation algorithm (and the page table walker in HW) to perform a bunch of extra loads in order to find whichever NAPOT PTE is first?  We wanted to stick with the address-translation algorithm as-is, in order to make it easier for implementations to adopt this idea.

Reply:
No extra loads, just mask off bits corresponding to the decoded napot_bits.

(In open-source implementations and discussions) I've seen ~3 ways to handle the A/D bits:

* Punt and always raise a page fault.  No change is needed.
* Do a page table walk using normal loads, then do a CAS (or LR/SC pair) to update the final PTE was fetched.  In this case we would only need to mask off bits corresponding to napot_bits.
* Do a page table walk using loads with LR semantics, then do a SC to update the PTE.  This case does require (exactly one) additional load, but it may be undesirable to begin with since it requires taking ownership of all PTEs, even if A=D=1 already.